File: src/main/java/com/datastruct/arraystack/scala/ArrayStack.scala
package com.datastruct.arraystack.scala

import ArrayStack.CAPACITY

class ArrayStack[T](capacity: Int = CAPACITY) extends Stack[T] {
  // 类初始化流程
  if (capacity <= 0) {
    throw new ExceptionStackEmpty("不能创建空栈，capacity值必须大于0")
  }

  protected val elementArray: Array[T] =  Array[Any](capacity).asInstanceOf[Array[T]] // 栈底层数组

  protected var topPointer: Int = -1  // 栈顶指针
  
  override def getSize: Int = topPointer + 1

  override def isEmpty: Boolean = topPointer == -1

  override def top: T = {
    if (topPointer == -1) {
      throw new ExceptionStackEmpty("栈为空，无法获取栈顶数据")
    }

    elementArray(topPointer)
  }

  override def push(element: T): Unit = {
    if (topPointer == capacity - 1) {
      throw new ExceptionArrayStackFull("栈已满，数据无法进栈")
    }

    topPointer += 1
    elementArray(topPointer)
  }

  override def pop: T = {
    if (topPointer == -1) {
      throw new ExceptionStackEmpty("栈为空，无数据可出栈")
    }

    topPointer -= 1
    elementArray(topPointer + 1)
  }
}

object ArrayStack {
  val CAPACITY = 1024 // 数组默认容量

  def apply[T](): ArrayStack[T] = new ArrayStack[T]()

  def apply[T](capacity: Int): ArrayStack[T] = new ArrayStack[T](capacity)
}

File: src/main/java/com/datastruct/arraystack/java/ArrayStack.java
package com.datastruct.arraystack.java;

public class ArrayStack<T> implements Stack<T> {
    public static final int CAPACITY = 1024;  // 数组默认容量
    protected int capacity; // 数组实际容量
    protected T[] arrayT;  // 对象数组
    protected int top = -1; // 栈顶元素指针

    public ArrayStack() {
        this(CAPACITY);
    }

    @SuppressWarnings("unchecked")
    public ArrayStack(int capacity) {
        if (capacity <= 0) {
            throw new ExceptionStackEmpty("不能创建空栈！容量必须大于0");
        }
        this.capacity = capacity;
        arrayT = (T[]) new Object[capacity];
    }

    @Override
    public int getSize() {
        return top + 1;
    }

    @Override
    public boolean isEmpty() {
        return top == -1;
    }

    @Override
    public T top() throws ExceptionStackEmpty {
        if (top == -1) {
            throw new ExceptionStackEmpty("栈为空，无法获取栈顶数据");
        }

        return this.arrayT[top];
    }

    @Override
    public void push(T t) {
        if (top == capacity - 1) {
            throw new ExceptionStackEmpty("栈已满，数据无法入栈");
        }
        ++top;
        arrayT[top] = t;
    }

    @Override
    public T pop() throws ExceptionStackEmpty {
        if (top == -1) {
            throw new ExceptionStackEmpty("栈为空，无数据可出栈");
        }
        --top;
        return arrayT[top + 1];
    }
}
